### PR TITLE
Room introduction & migration of recent search

### LINKED ISSUE #3111: Support recent search tracking with Room
Should be done:
* A kind of generic and abstract accessor to access Room data
* Secure the Room DB is created properly at start (if it is not easy to create on the fly) if necessary (first installation or migration context)
* The data structure and accessor to the recent searches (based on what exists in ObjectBox)
* Link with the existing code around the search UI (disconnect this from ObjectBox)
* Write the automated tests (Unit tests and maybe one UI test) for that new Room code (adapt what exists for ObjectBox)
* Write the migration function (for the data) from the ObjectBox recent_searches DB to the Room DB
* Write automated tests for that migration function

Once this is done #3115 will have to be implemented to secure we can test the migration.

RecentSearchEntity (Responisble for recent searches) https://github.com/kiwix/kiwix-android/blob/e2c901bf22034ea891c29ce12bc5793df7eccf52/core/src/main/java/org/kiwix/kiwixmobile/core/dao/entities/RecentSearchEntity.kt

### PR DESCRIPTION
**Pull request by gouri-panda**
Fixes #3111 

- [x]  A kind of generic and abstract accessor to access Room data
- [x] Secure the Room DB is created properly at start (if it is not easy to create on the fly) if necessary (first installation or migration context)
- [x] The data structure and accessor to the recent searches (based on what exists in ObjectBox)
- [x] Link with the existing code around the search UI (disconnect this from ObjectBox)
- [x] Write the automated tests (Unit tests and maybe one UI test) for that new Room code (adapt what exists for ObjectBox)
- [x] Write the migration function (for the data) from the ObjectBox recent_searches DB to the Room DB
- [x] Write automated tests for that migration function

### CODE REVIEW

**Comment by gouri-panda** [comment]
@MohitMaliDeveloper You can start reviewing this :)

File: core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            org.kiwix.kiwixmobile.core.search.adapter.SearchListItem.RecentSearchListItem(
+              searchEntity.searchTerm
+            )
+          }
+      }
+    }
+  }
+
+  @Query("INSERT INTO RecentSearchRoomEntity(searchTerm, zimId) VALUES (:title , :id)")
+  abstract fun saveSearch(title: String, id: String)
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt:67]: it should be zimId

Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt:67]: parameter name should be zimId.

File: core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  @Query("DELETE FROM RecentSearchRoomEntity")
+  abstract fun deleteSearchHistory()
+
+  fun migrationToRoomInsert(
+    box: Box<RecentSearchEntity>
+  ) {
+    val searchRoomEntityList = box.all
+    searchRoomEntityList.forEach {
+      CoroutineScope(Dispatchers.IO).launch {
+        saveSearch(it.searchTerm, it.id.toString())
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt:81]: hi @gouri-panda , 
currently previous recent search is not visible to user. because for now we are saving ```primary key id``` not the ```zim id```. so it should be replace by ```it.zimId```.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/Repository.kt [hunk 3]
```diff
@@ -28,6 +28,7 @@ import org.kiwix.kiwixmobile.core.dao.NewBookmarksDao
 import org.kiwix.kiwixmobile.core.dao.NewLanguagesDao
 import org.kiwix.kiwixmobile.core.dao.NewNoteDao
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/Repository.kt:30]: this is unused import we should remove this

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/modules/ApplicationModule.kt [hunk 4]
```diff
@@ -23,12 +23,14 @@ import android.app.NotificationManager
...
 import android.net.ConnectivityManager
 import android.os.storage.StorageManager
+import androidx.room.Room
 import dagger.Module
 import dagger.Provides
 import io.reactivex.Scheduler
 import io.reactivex.android.schedulers.AndroidSchedulers
 import io.reactivex.schedulers.Schedulers
 import org.kiwix.kiwixmobile.core.NightModeConfig
+import org.kiwix.kiwixmobile.core.data.local.KiwixRoomDatabase
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/di/modules/ApplicationModule.kt:33]: here too

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModel.kt [hunk 5]
```diff
@@ -35,6 +35,7 @@ import kotlinx.coroutines.launch
 import org.kiwix.kiwixmobile.core.R
 import org.kiwix.kiwixmobile.core.base.SideEffect
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModel.kt:37]: here too

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/DeleteRecentSearch.kt [hunk 6]
```diff
@@ -21,11 +21,12 @@ package org.kiwix.kiwixmobile.core.search.viewmodel.effects
 import androidx.appcompat.app.AppCompatActivity
 import org.kiwix.kiwixmobile.core.base.SideEffect
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/DeleteRecentSearch.kt:23]: here too.

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecents.kt [hunk 7]
```diff
@@ -19,16 +19,25 @@
...
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
+import org.kiwix.kiwixmobile.core.dao.NewRecentSearchRoomDao
 import org.kiwix.kiwixmobile.core.search.adapter.SearchListItem
 
 data class SaveSearchToRecents(
-  private val recentSearchDao: NewRecentSearchDao,
+  private val recentSearchDao: NewRecentSearchRoomDao,
   private val searchListItem: SearchListItem,
-  private val id: String?
+  private val id: String?,
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecents.kt:33]: it should be ```zimId```

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz Thanks for the review! It's not done yet but I'll add changes after it completes.

**Comment by kelson42** [comment]
What is the status here? Once the aitomated tests are done, I would like to start to play wirh the feature.

**Comment by gouri-panda** [comment]
@kelson42  Almost done. Once this has done we can move forward pretty quickly :)

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz @kelson42 It's Ready now.

**Comment by MohitMaliFtechiz** [comment]
Hi @gouri-panda ,
Ci is failing on unit coverage.

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz  I tried several times i don't know why :(

**Comment by MohitMaliFtechiz** [comment]
Hi @gouri-panda ,

* The error is in ```SearchViewModelTest.kt``` on line number 158. ```SaveSearchToRecents(recentSearchDao, searchListItem, "id")``` you forgot to pass ```viewmodelscope``` in this function.

* Second error is on line number 192. ```DeleteRecentSearch(searchListItem, recentSearchDao)``` in this function you forgot to pass ```viewmodelscope```.

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz Thanks! I forgot to check. Silly me :)

File: core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  @Query("INSERT INTO RecentSearchRoomEntity(searchTerm, zimId) VALUES (:title , :zimId)")
+  abstract fun saveSearch(title: String, zimId: String)
+
+  @Query("DELETE FROM RecentSearchRoomEntity WHERE searchTerm=:searchTerm")
+  abstract fun deleteSearchString(searchTerm: String)
+
+  @Query("DELETE FROM RecentSearchRoomEntity")
+  abstract fun deleteSearchHistory()
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewRecentSearchRoomDao.kt:73]: This function is not working. Object still persist in room database.

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/modules/ApplicationModule.kt [hunk 9]
```diff
@@ -63,6 +63,7 @@ class ApplicationModule {
 
   @Provides
   @Singleton
+
```
Comment (MohitMaliFtechiz) [inline core/src/main/java/org/kiwix/kiwixmobile/core/di/modules/ApplicationModule.kt:66]: Why we adding extra line here?

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

**Comment by MohitMaliFtechiz** [comment]
Hi @gouri-panda ,

We should write a test for ```deleteSearchHistory()``` as we wrote for other methods.

**Comment by gouri-panda** [comment]
> We should write a test for deleteSearchHistory() as we wrote for other methods.

@MohitMaliFtechiz It's ready to re-review now. We have the test for the delete. The bug was we were doing this on the main thread so it was not querying on the database. In the test we are testing on the main thread by default so we didn't notice that before.

**Review by MohitMaliFtechiz — APPROVED** [review]

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz Thanks for the review :)

**Comment by kelson42** [comment]
@gouri-panda Do I have an APK somewhere to test this PR?

**Comment by gouri-panda** [comment]
@kelson42 It's not in the nightly yet. I'll give it you.

**Review by MohitMaliFtechiz — APPROVED** [review]

**Review by kelson42 — CHANGES_REQUESTED** [review]
Few remarks:
* I have tested and it seems to work
* On a high level, no big problems

Here a few things to change:
* Not really happy with LIcense header, you should have the main author email and at the bottom a better link to the license (I will try to find something better)
* Please upgrade your tests to test the edge case (undefined values, empty values, really long values, values with strange unicode characters)
* The migration should happen at start of the app once (I guess this is how it has been done from the squidb/ObjectBox migration). All the further operations should assume the RoomDB is available and ready to use. Reason is (at least) testing something once is more efficient that N times. On the top of this, if the migration takes a few seconds, it is better to wait once at start than during any other operation having a UI impact (like searching).

**Comment by kelson42** [comment]
@gouri-panda Any update on this PR?

**Comment by gouri-panda** [comment]
@kelson42 Done :)

**Comment by MohitMaliFtechiz** [comment]
@gouri-panda , CI is failing.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@gouri-panda , I have few questions.

> * Please upgrade your tests to test the edge case (undefined values, empty values, really long values, values with strange unicode characters)

> * The migration should happen at start of the app once (I guess this is how it has been done from the squidb/ObjectBox migration). All the further operations should assume the RoomDB is available and ready to use. Reason is (at least) testing something once is more efficient that N times. On the top of this, if the migration takes a few seconds, it is better to wait once at start than during any other operation having a UI impact (like searching).

* Have you cover these points as @kelson42 said ? I'm not seeing any changes related to these points or am i missing something?

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  companion object {
+    private var db: KiwixRoomDatabase? = null
+    fun getInstance(context: Context, boxStore: BoxStore): KiwixRoomDatabase {
+      return db ?: synchronized(KiwixRoomDatabase::class) {
+        return@getInstance db
+          ?: Room.databaseBuilder(context, KiwixRoomDatabase::class.java, "KiwixRoom.db")
+            // We have already database name called kiwix.db in order to avoid complexity we named as
+            // kiwixRoom.db
+            .build().also {
+              if (!BuildConfig.BUILD_TYPE.contentEquals("fdroid")) {
```
[the remark this answers is not in the document]
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: @MohitMaliFtechiz We aren't migrating the data in fdroid flavour.

File: app/build.gradle.kts [hunk 11]
```diff
@@ -56,24 +56,31 @@ android {
...
+      matchingFallbacks += "baseRelease"
+    }
     create("playStore") {
       manifestPlaceholders += mapOf()
       initWith(getByName("release"))
       matchingFallbacks += "release"
       buildConfigField("boolean", "IS_PLAYSTORE", "true")
       manifestPlaceholders["permission"] = "android.permission.placeholder"
     }
-    create("nightly") {
```
[the remark this answers is not in the document]
Comment (gouri-panda) [reply app/build.gradle.kts:74]: Thanks!  I forgot to add this while refactoring.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: So you meant we are totally remove the object box dependency from fdroid build?

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz I have discussed with @kelson42 about this. The database is already migrating the data while it installs the database the first time and I'll add the edge cases test cases when all the pending pr will be merged because it would be easier to do it then.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (kelson42) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: Yes, we have too, otherwise we won't be able anymore to publish on FDroid. We need to be able to compile:
* With ObjectBox for PlayStore (during the time we want to allow user to migrate their data to Room)
* Without ObjectBox in all other scenarios.

We need therefore a special compilation target for Google in the Gradle

[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: * But we are initialise fdroid from ```baseRelease``` and baseRelease from ```debug``` .
* For debug we are adding object box dependency in core module.

`
    getByName("debug") {
      dependencies {
        implementation(Libs.objectbox_kotlin)
        implementation(Libs.objectbox_rxjava)
      }
    }
`

so it will automatically added in fdroid module as well.

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz I have discussed with @kelson42 about this. The database is already migrating the data while it installs the database the first time and I'll add the edge cases test cases when all the pending pr will be merged because it would be easier to do it then.

Thanks for clarification.

File: core/build.gradle.kts [hunk 12]
```diff
@@ -47,8 +47,26 @@ android {
...
     getByName("release") {
       isMinifyEnabled = false
+      apply(plugin = "io.objectbox")
+      dependencies {
+        implementation(Libs.objectbox_kotlin)
+        implementation(Libs.objectbox_rxjava)
+      }
+    }
+    create("playStore") {
+      apply(plugin = "io.objectbox")
```
Comment (MohitMaliFtechiz) [inline core/build.gradle.kts:65]: here too.

File: core/build.gradle.kts [hunk 13]
```diff
@@ -47,8 +47,26 @@ android {
...
   buildTypes {
+    getByName("debug") {
+      dependencies {
+        implementation(Libs.objectbox_kotlin)
+        implementation(Libs.objectbox_rxjava)
+      }
+    }
     getByName("release") {
       isMinifyEnabled = false
+      apply(plugin = "io.objectbox")
```
Comment (MohitMaliFtechiz) [inline core/build.gradle.kts:58]: why we are adding this code here? we already declared this here https://github.com/kiwix/kiwix-android/blob/563170cd76c9b6f2e8a0562dbefd89850e551249/core/build.gradle.kts#L19

File: app/build.gradle.kts [hunk 14]
```diff
@@ -75,6 +81,11 @@ android {
       initWith(getByName("debug"))
       setMatchingFallbacks("debug")
```
Comment (MohitMaliFtechiz) [inline app/build.gradle.kts:82]: This method is deprecated. please change this with ```matchingFallbacks += "debug"``` .

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  @Test
+  @Throws(IOException::class)
+  fun testMigrationTest() = runBlocking {
+    val context = ApplicationProvider.getApplicationContext<Context>()
+    db = Room.inMemoryDatabaseBuilder(
+      context, KiwixRoomDatabase::class.java
+    ).build()
+    recentSearchRoomDao = db.recentSearchRoomDao()
+    val searchTerm = "title"
+    val zimId = "zimId"
```
Comment (kelson42) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt:51]: Not sure what is `zimID`, but if this is really a ZIM internal ID, then please use a realistic UUID one. Everywhere.

**Review by kelson42 — CHANGES_REQUESTED** [review]
* Please split in different classes Room accessor code and migration code from ObjectBox to Room. This is two different things. Same for the tests classes.
* `KiwixRoomDatabaseTest.kt` is a migration test and should be called like this.
* In `KiwixRoomDatabaseTest.kt`, making 3 tests almost similar is not really valuable, you can test the same in only one test. In particular you should test the edge cases (empty values, values with unicode charaters, undefined values, etc...). This remark should be apply for all automated tests written in this PR.
* What are the maximal sizes of values and how the library deal with it? Should be tested as well.
* For `RecentSearchRoomDaoTest.kt`, beside the remarks on the above comment which apply as well, the overall call can be significantly reduced. Pretty sure you can test better in less than 60 lines.
* Migration should done once at start if necessary. All the rest of the code should assume it is done. That way is clearer and will run faster.

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: @MohitMaliFtechiz Done https://github.com/kiwix/kiwix-android/pull/3143/commits/62f8378a345885b4dc680adb13a33f381f0c0e47

File: core/build.gradle.kts (hunk 12 above)
Comment (gouri-panda) [reply core/build.gradle.kts:65]: @MohitMaliFtechiz those are for the different build types for different stores.

**Comment by gouri-panda** [comment]
> Please split in different classes Room accessor code and migration code from ObjectBox to Room. This is two different things. Same for the tests classes.

@kelson42 Gradle will automatically remove the library dependencies and code as well.


>In KiwixRoomDatabaseTest.kt, making 3 tests almost similar is not really valuable, you can test the same in only one test. In particular you should test the edge cases (empty values, values with unicode charaters, undefined values, etc...). This remark should be apply for all automated tests written in this PR.

We haven't done much there to test. I've added more test as well for now. If i think i can add more tests , I'll add later. We'll add more tests in use cases.

>What are the maximal sizes of values and how the library deal with it? Should be tested as well.

I don't understand the question. Can you clarify a little more?

>Migration should done once at start if necessary. All the rest of the code should assume it is done. That way is clearer and will run faster.

We are already doing it here :)

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: @MohitMaliFtechiz I think it's a regression. See https://github.com/kiwix/kiwix-android/issues/3282. After we fix it we can achieve it.

[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: @gouri-panda , i think this is import error like it's ```androidx.multidex.BuildConfig``` not ```org.kiwix.kiwixmobile.BuildConfig``` if we are using the ```org.kiwix.kiwixmobile.BuildConfig``` then we can get every build variants like in below image or if we want to get build type from ```androidx.multidex.BuildConfig```, so i think first we needs to set build type on ```androidx.multidex.BuildConfig```.

Screenshot from 2023-03-31 17-48-45

File: core/build.gradle.kts (hunk 12 above)
Comment (MohitMaliFtechiz) [reply core/build.gradle.kts:65]: @gouri-panda Okay .

* But we are adding this plugin on top of ```core/build.gradle``` (By this way it's include this plugin in every variant). https://github.com/kiwix/kiwix-android/blob/563170cd76c9b6f2e8a0562dbefd89850e551249/core/build.gradle.kts#L19
So why we are adding this again in ```release``` , ```playstore``` variant?

* We are removing the ```io.objectbox``` dependency from fdroid build, so is it necessary to add ```apply(plugin = "io.objectbox")``` in ```release``` , ```playstore``` variants?
```
"fdroidImplementation"(
    Libs.objectbox_kotlin,
  ) {
    exclude("io.objectbox")
  }
```

Comment (gouri-panda) [reply core/build.gradle.kts:64]: @MohitMaliFtechiz Done https://github.com/kiwix/kiwix-android/pull/3143/commits/2b95ad2bf155b2a9a781115dd07a795fad904cfa. 

>We are removing the io.objectbox dependency from fdroid build, so is it necessary to add apply(plugin = "io.objectbox") in release , playstore variants?

Yes!, in order to use the Objectbox dependency we should add the plugin as well.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@gouri-panda just need small changes, then we are good to move forward.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    val context = ApplicationProvider.getApplicationContext<Context>()
+    db = Room.inMemoryDatabaseBuilder(
+      context, KiwixRoomDatabase::class.java
+    ).build()
+    recentSearchRoomDao = db.recentSearchRoomDao()
+    val searchTerm = "title"
+    val zimId = "8812214350305159407L"
+    box.put(RecentSearchEntity(searchTerm = searchTerm, zimId = zimId))
+    recentSearchRoomDao.migrationToRoomInsert(box)
+    recentSearchRoomDao.search("zimId").collect { recentSearchEntites ->
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt:54]: IMO, Here we should use val zimId = "8812214350305159407L" this zimId instead of string "zimId" like we are using in RecentSearchRoomDaoTest . Because we are saving the zimId in room database, so we should fetch the same id from database.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+      val entity = recentSearchEntites.find { it.zimId == zimId }
+      if (entity != null) {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt:56]: Why we are adding this condition here. Because IMO, We are saving the data in room in above statement, so it should not be null here.

File: buildSrc/src/main/kotlin/Versions.kt [hunk 18]
```diff
@@ -103,6 +103,16 @@ object Versions {
...
+
+  const val roomVersion = "2.3.0"
+
+
+  /**
+   * Current version: "6.2"
+   * See issue 19: How to update Gradle itself?
+   * https://github.com/jmfayard/buildSrcVersions/issues/19
+   */
+  const val gradleLatestVersion: String = "6.7.1"
```
Comment (MohitMaliFtechiz) [inline buildSrc/src/main/kotlin/Versions.kt:115]: This code is unused here, we already removed this code in https://github.com/kiwix/kiwix-android/pull/3278/files . Please remove this code.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt (hunk 17 above)
Comment (gouri-panda) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabaseTest.kt:56]: >      val entity = recentSearchEntites.find { it.zimId == zimId }

@MohitMaliFtechiz We are adding this data into database and here we finds it. if it in case it doesn't find it then we are preventing from the Null pointer Exception.

File: buildSrc/src/main/kotlin/Versions.kt (hunk 18 above)
Comment (gouri-panda) [reply buildSrc/src/main/kotlin/Versions.kt]: From the rebase we get it! Done https://github.com/kiwix/kiwix-android/pull/3143/commits/30a04b8cb21f6abbdc22a040aa781affd6d6e5d2

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: Hi @gouri-panda , I think you forgot to change the import here.

[the remark this answers is not in the document]
Comment (MohitMaliFtechiz) [reply core/src/main/java/org/kiwix/kiwixmobile/core/data/local/KiwixRoomDatabase.kt:45]: I have did changes.

**Review by MohitMaliFtechiz — APPROVED** [review]
@gouri-panda Thanks for your hard work :tada: :tada::tada: , Now it's LGTM.

**Comment by kelson42** [comment]
> > Please split in different classes Room accessor code and migration code from ObjectBox to Room. This is two different things. Same for the tests classes.
> 
> @kelson42 Gradle will automatically remove the library dependencies and code as well.

I insist, the code for the migration has its own duty and properties. It has nothing to do with the Room accessor itself. This is a software architecture problem and Gradle has nothing to do with it.

> > In KiwixRoomDatabaseTest.kt, making 3 tests almost similar is not really valuable, you can test the same in only one test. In particular you should test the edge cases (empty values, values with unicode charaters, undefined values, etc...). This remark should be apply for all automated tests written in this PR.
> 
> We haven't done much there to test. I've added more test as well for now. If i think i can add more tests , I'll add later. We'll add more tests in use cases.

I have checked again and almost nothing has changed. I have no way to verify that the code handles properly unicode ... and your test code has a lot of heavy redundancies.

> > What are the maximal sizes of values and how the library deal with it? Should be tested as well.
> 
> I don't understand the question. Can you clarify a little more?

Can i put a search term of 2MB?

> > Migration should done once at start if necessary. All the rest of the code should assume it is done. That way is clearer and will run faster.
> 
> We are already doing it here :)

Not, it's not AFAIK. It's done https://github.com/kiwix/kiwix-android/pull/3143/files#diff-f3b99b0edaf474b49786f9a0a4ce2d9c354ebe51bde241c1637d2e22d6cf03cdR37 which is basically tested at each instanciation of `KiwixRoomDatabase`.
